### DISCUSSION TITLE
How I can use it from C++ code?

### DISCUSSION THREAD

**Discussion by AndreyAlifanov** [tarantool/queue#193] [Q&A] [answered]
Hello!

How I can use these queues from C++ code?

**Comment by Totktonada** [comment, marked as answer]
It depends of a nature of this C++ code.

If we're talking about an external process, one should choose a connector and call queue functions using it. Don't forget to [grant](https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_schema/user_grant/) privileges to execute those functions to your user/role, consider using `queue.grant({call = true})`. You can look at the list of C/C++ connectors [here](https://www.tarantool.io/en/download/connectors). At the moment of writting it contains [tarantool-c](https://github.com/tarantool/tarantool-c) (C) and [tntcxx](https://github.com/tarantool/tntcxx) (C++).

If we're about a module (Lua rock) or a [stored procedure](https://www.tarantool.io/en/doc/latest/how-to/app/c_tutorial/) (registered with [`box.schema.func.create()`](https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_schema/func_create/)) written on C/C++, things are a bit different. We want to call a queue function locally. [Lua/C API](https://www.lua.org/manual/5.1/manual.html#3) is for that.

A sketchy code:

```c
int saved_top = lua_gettop(L);

lua_getfield(L, LUA_GLOBALSINDEX, "queue");
lua_getfield(L, -1, "take");
<..prepare queue.take() args..>
if (lua_pcall(L, <...>) != 0) {
    <..handle an error..>
}
<..process return values..>

lua_settop(L, saved_top);
```

If we're in a stored procedure, we have no `L` (Lua stack). We should create our own Lua stack using `lua_newthread()` (with `luaT_state()` as an argument) and keep a reference to it using `luaL_ref()`. We can't use global Lua stack (`luaT_state()`) directly, because it shouldn't be changed in a way that is observable for other code (IOW, if we call something or yield current fiber).

```c
static struct lua_State *lua_state = NULL;
static int lua_state_ref = LUA_NOREF;

/* Create a Lua stack. */
struct lua_State *global_L = luaT_state();
lua_state = lua_newthread(global_L); /* Popped by luaL_ref(). */
/* We should keep a reference to protect it from GC. */
lua_state_ref = luaL_ref(global_L, LUA_REGISTRYINDEX);

/* Release the Lua stack. */
struct lua_State *global_L = luaT_state();
luaL_unref(global_L, LUA_REGISTRYINDEX, lua_state_ref);
lua_state = NULL;
lua_state_ref = LUA_NOREF;
```

Use this `lua_state` as `L` in your functions.

All the code snippets are just to illustrate the idea and show APIs that can be used here. I didn't perform any testing to offer a ready-to-use solution.

**Reply by AndreyAlifanov** [reply]
Thank you for information.
I want to use queues from external code. I don't want to write modules or procedures.

I've tried to write simple code with tntcxx, but it is very fragile and buggy.
Because it works through Connection.call only, and I've got errors from Lua.
So, I want to find out more info about API, to write these APIs on my own.

**Reply by AndreyAlifanov** [reply]
I mean calls like queue.create_tube and so on.
As I guess, they should work with Connection.eval, but it's absent in tntcxx.

**Reply by Totktonada** [reply]
I suggest to prefer `call` over `eval`, because if you'll allow a user to execute `eval`, you'll allow everything. If you'll later want to restrict the user, it may be tiresome: a lot of already written code requires some privileges, but they are not granted explicitly. At this point one need to inspect/test all the code and grant necessary privileges.

If you want a help with particular errors, let's share them. I can't help without any information. I suspect that it is about privileges and you need to look into direction of `box.schema.user.grant()` or `queue.grant({call = true})`.

It is better to ask questions regarding tntcxx in its namespace (say, in issues). Unlikely its maintainers consistently look into queue discussions/issues. If you'll do, let's provide them a simple reproducer of your problem: without queue, just define some Lua function by yourself. It'll increase probability that your problem will gain enough attention.

**Reply by AndreyAlifanov** [reply]
User will not have access to this functionality. It will be inside server-side code.
But I see your point of view.
Thank you for help.
I'll try to ask in tntcxx branch, with example.

**Reply by Totktonada** [reply]
For the record, the tntcxx related discussion is here: https://github.com/tarantool/tntcxx/issues/33.
